gdpr/models.py

```diff
@@ -11,7 +12,64 @@
 
 class LegalReasonManager(models.Manager):
 
-    def create_from_purpose_slug(self, purpose_slug, source_object, issued_at=None, tag=None, related_objects=None):
+    def create_consent(self, purpose_slug, source_object, issued_at=None, tag=None, related_objects=None):
+        """
+        Create (or update, if it exist) a LegalReason with purpose slug for concrete object instance
+
+        Args:
+            purpose_slug: String of Legal Reason purpose
+            source_object: Source object this Legal Reason is related to
+            issued_at: When the Legal Reason consent was given
+            tag:
+            related_objects: Objects this Legal Reason relates to (ie. order, registrations etc.)
+
+        Returns:
+            Legal Reason: LegalReason object
+        """
+        if self.filter_source_instance(source_object).filter(purpose_slug=purpose_slug).exists():
```

Comment:
proč jsi tu vlastně nepoužil create_or_update?

Reply:
měl jsi k tomu nějaký důvod?

Reply:
Jestli dobře vidím, create_or_update je v pystonu a k něčemu jinému (tuhle závislost nemáme) - můžu zkusit použít update_or_create z Django, pak ale budu muset zvednout minimální požadovanou verzi Django (měli bychom se do ní vejít), pokud to vůbec bude fungovat.

Reply:
já myslel z djanga, jen sjem otočil update a create v té metodě. Jak tě napadl pyston?

Reply:
a django 1.11 update_or_create už má

Reply:
1.10 a níže už ignoruj

Reply:
Ale stejně

> já myslel z djanga, jen sjem otočil update a create v té metodě. Jak tě napadl pyston?

hledal jsem "create_or_update"

> a django 1.11 update_or_create už má

JJ, všiml jsem si - ale stejně mě nenapadá jak to elegantně vyřešit pomocí update_or_create, pokud nechceme přepsat `issued_at` - což byl, pokud se dobře pamatuji, požadavek zachovat (datum prvotního udělení souhlasu, před updatem).
Tak jsem to alespoň předělal do transakce typu "zkus aktualizovat a pokud není co, tak vytvoř".

Lepší?

Reply:
v tom případě to budeš muset udělat v bloku https://docs.djangoproject.com/en/2.1/ref/models/querysets/#select-for-update atomicky

Reply:
uděláš to i bez toho přes get_or_create

Reply:
nejdřív objekt získáš/nebo vytvořís default hodnotami a pak pokud nebyl vytvořený tak mu upravíš to co potřebuješ